source/solvers/mf_navier_stokes_operators.cc

```diff
@@ -877,3 +903,394 @@ NavierStokesSUPGPSPGOperator<dim, number>::local_evaluate_residual(
+
+      // Gather the original value/gradient
+      typename FECellIntegrator::value_type    value = integrator.get_value(q);
+      typename FECellIntegrator::gradient_type gradient =
+        integrator.get_gradient(q);
+      typename FECellIntegrator::gradient_type hessian_diagonal =
+        integrator.get_hessian_diagonal(q);
+
+      // Result value/gradient we will use
+      typename FECellIntegrator::value_type    value_result;
+      typename FECellIntegrator::gradient_type gradient_result;
+
+      // Gather previous values of the velocity and the pressure
+      auto previous_values   = this->nonlinear_previous_values(cell, q);
+      auto previous_gradient = this->nonlinear_previous_gradient(cell, q);
+      auto previous_hessian_diagonal =
+        this->nonlinear_previous_hessian_diagonal(cell, q);
+
+
+      // Time derivatives of previous solutions
+      auto previous_time_derivatives =
+        this->time_derivatives_previous_solutions(cell, q);
+
+      // Calculate tau
+      VectorizedArray<number> u_mag = VectorizedArray<number>(1e-12);
+      VectorizedArray<number> tau   = VectorizedArray<number>(0.0);
+
+      std::array<number, VectorizedArray<number>::size()> h_k;
+      std::array<number, VectorizedArray<number>::size()> h;
+
+      for (auto lane = 0u;
+           lane < this->matrix_free.n_active_entries_per_cell_batch(cell);
+           lane++)
+        {
+          h_k[lane] =
+            this->matrix_free.get_cell_iterator(cell, lane)->measure();
+        }
+
+      for (unsigned int v = 0; v < VectorizedArray<number>::size(); ++v)
+        {
+          if (dim == 2)
+            {
+              h[v] = std::sqrt(4. * h_k[v] / M_PI) / this->fe_degree;
+            }
+          else if (dim == 3)
+            {
+              h[v] = std::pow(6 * h_k[v] / M_PI, 1. / 3.) / this->fe_degree;
+            }
+        }
+
+      for (unsigned int k = 0; k < dim; ++k)
+        u_mag += Utilities::fixed_power<2>(previous_values[k]);
+
+      for (unsigned int v = 0; v < VectorizedArray<number>::size(); ++v)
+        {
+          tau[v] = 1. / std::sqrt(
+                          Utilities::fixed_power<2>(sdt) +
+                          4. * u_mag[v] / h[v] / h[v] +
+                          9 * Utilities::fixed_power<2>(
+                                4 * this->kinematic_viscosity / (h[v] * h[v])));
+        }
+
+      // Weak form jacobian
+      for (unsigned int i = 0; i < dim; ++i)
+        {
+          // ν(∇v,∇δu)
+          gradient_result[i] = this->kinematic_viscosity * gradient[i];
+          // -(∇·v,δp)
+          gradient_result[i][i] += -value[dim];
+          // +(q,∇δu)
+          value_result[dim] += gradient[i][i];
+
+          for (unsigned int k = 0; k < dim; ++k)
+            {
+              // +(v,(u·∇)δu)
+              value_result[i] += gradient[i][k] * previous_values[k];
+              // +(v,(δu·∇)u)
+              value_result[i] += previous_gradient[i][k] * value[k];
```

Comment:
I guess in the formula and implementation the terms are swapped.

Reply:
Yes, they are swapped.